### ISSUE TITLE
Should Sensor release() be a no-op if called multiple times

### ISSUE THREAD

**Issue by alexpareto** [expo/expo#6204]
Currently, in Android - the `release()` function for sensor subscriptions calls `assertSubscriptionIsAlive`. This method, `assertSubscriptionIsAlive` will cause a fatal crash if the subscription has already been released. Is this necessary? 

Would it make more sense for Android to not go through a fatal crash, and just be harmless no-op if a subscription is released twice? From a quick glance through, I believe this is the behavior in iOS - where there is no crash if a subscription is attempted to be removed twice. This would also explain why this crash occurs mainly on Android devices.

Here is the Android Code:
https://github.com/expo/expo/blob/8d441c7e9b081cfdd6a45edc165251d7273225a0/packages/expo-sensors/android/src/main/java/expo/modules/sensors/services/SensorServiceSubscription.java#L52-L62

This is the iOS Code:
https://github.com/expo/expo/blob/891b93f84d03c775c371fa9d7240b3792f8a68e2/ios/Exponent/Kernel/Services/EXSensorManager.m#L81-L87

**Comment by cruzach** [comment]
Maybe could we replace `assertSubscriptionIsAlive` with `if (!mHasBeenReleased)`?

Can you provide a snack that can repro the crash?

**Comment by alexpareto** [comment]
Hard to provide a Snack that can reproduce, as it appears to be related to the app going into the background, then being pulled out of the background. Here is the crash report:
```
java.lang.IllegalStateException: Subscription has been released, cannot call methods on a released subscription.
    at expo.modules.sensors.services.SensorServiceSubscription.assertSubscriptionIsAlive(SensorServiceSubscription.java:60)
    at expo.modules.sensors.services.SensorServiceSubscription.stop(SensorServiceSubscription.java:45)
    at expo.modules.sensors.modules.BaseSensorModule.onHostPause(BaseSensorModule.java:102)
    at org.unimodules.adapters.react.services.UIManagerModuleWrapper$3.onHostPause(UIManagerModuleWrapper.java:162)
    at com.facebook.react.bridge.ReactContext.onHostPause(ReactContext.java:236)
    at com.facebook.react.ReactInstanceManager.moveToBeforeResumeLifecycleState(ReactInstanceManager.java:674)
    at com.facebook.react.ReactInstanceManager.onHostPause(ReactInstanceManager.java:498)
    at com.facebook.react.ReactInstanceManager.onHostPause(ReactInstanceManager.java:517)
    at com.facebook.react.ReactActivityDelegate.onPause(ReactActivityDelegate.java:99)
    at com.facebook.react.ReactActivity.onPause(ReactActivity.java:58)
    at android.app.Activity.performPause(Activity.java:7167)
    at android.app.Instrumentation.callActivityOnPause(Instrumentation.java:1408)
    at android.app.ActivityThread.performPauseActivityIfNeeded(ActivityThread.java:3942)
    at android.app.ActivityThread.performPauseActivity(ActivityThread.java:3919)
    at android.app.ActivityThread.performPauseActivity(ActivityThread.java:3893)
    at android.app.ActivityThread.handlePauseActivity(ActivityThread.java:3867)
    at android.app.ActivityThread.-wrap15
    at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1643)
    at android.os.Handler.dispatchMessage(Handler.java:106)
    at android.os.Looper.loop(Looper.java:164)
    at android.app.ActivityThread.main(ActivityThread.java:6548)
    at java.lang.reflect.Method.invoke(Method.java)
    at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:438)
    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:857)
java.lang.RuntimeException: Unable to pause activity {app.name}: java.lang.IllegalStateException: Subscription has been released, cannot call methods on a released subscription.
    at android.app.ActivityThread.performPauseActivityIfNeeded(ActivityThread.java:3953)
    at android.app.ActivityThread.performPauseActivity(ActivityThread.java:3919)
    at android.app.ActivityThread.performPauseActivity(ActivityThread.java:3893)
    at android.app.ActivityThread.handlePauseActivity(ActivityThread.java:3867)
    at android.app.ActivityThread.-wrap15
    at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1643)
    at android.os.Handler.dispatchMessage(Handler.java:106)
    at android.os.Looper.loop(Looper.java:164)
    at android.app.ActivityThread.main(ActivityThread.java:6548)
    at java.lang.reflect.Method.invoke(Method.java)
    at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:438)
    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:857)
```

Looking at this crash report again, the culprit actually seems to be the `onHostPause` in `BaseSensorModule.java`  and the `stop()` method in SensorServiceSubscription (not the release() function as initially thought) - which is called twice (once when the app is backgrounded and again when the app is resumed). Since onPause calls stop() without any checks, the Subscription is attempted to be cleared twice:
https://github.com/expo/expo/blob/1d5998d7b12e09194c889abb361f51d283da5083/packages/expo-sensors/android/src/main/java/expo/modules/sensors/modules/BaseSensorModule.java#L100-L103

Stop method:
https://github.com/expo/expo/blob/8d441c7e9b081cfdd6a45edc165251d7273225a0/packages/expo-sensors/android/src/main/java/expo/modules/sensors/services/SensorServiceSubscription.java#L44-L50

As described in the React Native docs, `onPause` is called when the app is backgrounded, but also when the app is resumed. The Subscription in Expo is being cleared initially when the app is backgrounded (onPause) and then again, when the app is resumed (onPause):
https://github.com/facebook/react-native/blob/master/ReactAndroid/src/main/java/com/facebook/react/bridge/LifecycleEventListener.java#L10-L29

I think your solution makes sense - replace `assertSubscriptionIsAlive()` or do a verification in the `onHostPause` method that the subscription is still active, a new `onHostPause` could be:
```java
  @Override
  public void onHostPause() {
if (getSensorKernelServiceSubscription().isEnabled()) {
    getSensorKernelServiceSubscription().stop();
}
}
```

With that said, I lean towards your solution. It seems risky to trust the caller all the time to abide by that `assertSubscriptionIsAlive` statement and dramatic to fall back to a hard crash if they do not. Calling stop() or release() feels like it should be safe to call multiple times (reminds me of calling rows.Close() for db access in Go). 

What do you think?

**Comment by cruzach** [comment]
> Hard to provide a Snack that can reproduce, as it appears to be related to the app going into the background, then being pulled out of the background

that's fine, I can take the code from the snack and run it locally if need be

**Comment by ide** [comment]
I agree we shouldn’t crash and should just do nothing, or throw a JS exception. Regardless, this is part of the public API and should be consistent across all three underlying platforms Expo supports.

**Comment by tourze** [comment]
Same problem, I think this exception should not been thrown.
![image](https://user-images.githubusercontent.com/13899502/69003691-82c9aa00-0941-11ea-8f35-ec1a8c226c7b.png)

**Comment by tourze** [comment]
Is there any idea to solve it ?

**Comment by ide** [comment]
Unless there is a bug (and there might be) if you cleanly unsubscribe exactly as many times as you subscribe it sounds like you shouldn't get this error.

**Comment by alexpareto** [comment]
I would be supportive of moving the behavior to be consistent between iOS and Android. Would perhaps make these issues easier to debug & reproduce.

Android hard crashes while iOS silently fails. 

iOS:
https://github.com/expo/expo/blob/891b93f84d03c775c371fa9d7240b3792f8a68e2/ios/Exponent/Kernel/Services/EXSensorManager.m#L81-L87

Android:
expo/packages/expo-sensors/android/src/main/java/expo/modules/sensors/services/SensorServiceSubscription.java

**Comment by rodneyrehm** [comment]
This is still happening in SDK 40 Bare Workflow.

```
java.lang.IllegalStateException: Subscription has been released, cannot call methods on a released subscription.
    at expo.modules.sensors.services.SensorServiceSubscription.assertSubscriptionIsAlive(SensorServiceSubscription.java:60)
    at expo.modules.sensors.services.SensorServiceSubscription.start(SensorServiceSubscription.java:20)
    at expo.modules.sensors.modules.BaseSensorModule.onHostResume(BaseSensorModule.java:98)
    at org.unimodules.adapters.react.services.UIManagerModuleWrapper$3.onHostResume(UIManagerModuleWrapper.java:127)
    at com.facebook.react.bridge.ReactContext.onHostResume(ReactContext.java:240)
    at com.facebook.react.ReactInstanceManager.moveToResumedLifecycleState(ReactInstanceManager.java:715)
    at com.facebook.react.ReactInstanceManager.onHostResume(ReactInstanceManager.java:619)
    at com.facebook.react.ReactInstanceManager.onHostResume(ReactInstanceManager.java:576)
    at com.facebook.react.ReactDelegate.onHostResume(ReactDelegate.java:53)
    at com.facebook.react.ReactActivityDelegate.onResume(ReactActivityDelegate.java:99)
    at com.facebook.react.ReactActivity.onResume(ReactActivity.java:57)
    at android.app.Instrumentation.callActivityOnResume(Instrumentation.java:1308)
    at android.app.Activity.performResume(Activity.java:6067)
    at android.app.ActivityThread.performResumeActivity(ActivityThread.java:3322)
    at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:3374)
    at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1482)
    at android.os.Handler.dispatchMessage(Handler.java:106)
    at android.os.Looper.loop(Looper.java:209)
    at android.app.ActivityThread.main(ActivityThread.java:5900)
    at java.lang.reflect.Method.invoke(Method.java)
    at java.lang.reflect.Method.invoke(Method.java:372)
    at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1005)
    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:800)

java.lang.RuntimeException: Unable to resume activity {<redacted>/<redacted>.MainActivity}: java.lang.IllegalStateException: Subscription has been released, cannot call methods on a released subscription.
    at android.app.ActivityThread.performResumeActivity(ActivityThread.java:3343)
    at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:3374)
    at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1482)
    at android.os.Handler.dispatchMessage(Handler.java:106)
    at android.os.Looper.loop(Looper.java:209)
    at android.app.ActivityThread.main(ActivityThread.java:5900)
    at java.lang.reflect.Method.invoke(Method.java)
    at java.lang.reflect.Method.invoke(Method.java:372)
    at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1005)
    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:800)
```

**Comment by brentvatne** [comment]
a pr would be welcome

**Comment by ixf** [comment]
This was resolved in https://github.com/expo/expo/pull/6828/commits/07b838c83c93a41b2ab37567bf62ac12dc3ba687 (part of https://github.com/expo/expo/pull/6828 )

> This is still happening in SDK 40 Bare Workflow.
> 
> ```
> java.lang.IllegalStateException: Subscription has been released, cannot call methods on a released subscription.
>     at expo.modules.sensors.services.SensorServiceSubscription.assertSubscriptionIsAlive(SensorServiceSubscription.java:60)
>     at expo.modules.sensors.services.SensorServiceSubscription.start(SensorServiceSubscription.java:20)
>     ...
> ```

This seems to me like a different issue (calling SensorServiceSubscription.start after subscription has been released)

**Comment by fermuch** [comment]
> This was resolved in 07b838c (part of #6828 )

@ixf is the fix in a released version? I cannot find it.

> This seems to me like a different issue (calling SensorServiceSubscription.start after subscription has been released)

Has there been any progress since then?
